doc: consolidate info on manual linux kernel configs (#263056)
NixOS/nixpkgs: doc/packages/linux.section.md

The Nix expressions to build the Linux kernel are in [`pkgs/os-specific/linux/kernel`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/linux/kernel).

The function [`pkgs.buildLinux`](https://github.com/NixOS/nixpkgs/blob/f9d4f5f23ce7ccb7ad91d6d2150e367baf7014f6/pkgs/os-specific/linux/kernel/generic.nix) builds a kernel with [common configuration values](hthttps://github.com/NixOS/nixpkgs/blob/f9d4f5f23ce7ccb7ad91d6d2150e367baf7014f6/pkgs/os-specific/linux/kernel/common-config.nix).
This is the preferred option unless you have a very specific use case.
Most kernels packaged in Nixpkgs are built that way, and it will also generate kernels suitable for NixOS.

> [`pkgs.linuxManualConfig`](https://github.com/NixOS/nixpkgs/blob/f9d4f5f23ce7ccb7ad91d6d2150e367baf7014f6/pkgs/os-specific/linux/kernel/manual-config.nix) requires a complete configuration to be passed, and has less additional features than `pkgs.buildLinux`, such as the use of common configuration values and exposing the `features` attribute, as explained below.

Reviewer: less is for uncountable *amounts*, fewer is for countable *numbers*.

It's not clear to me which is which in terms of features. Pick the one that's correct:

```suggestion
[`pkgs.linuxManualConfig`](https://github.com/NixOS/nixpkgs/blob/f9d4f5f23ce7ccb7ad91d6d2150e367baf7014f6/pkgs/os-specific/linux/kernel/manual-config.nix) requires a complete configuration to be passed.
It has fewer additional features than `pkgs.buildLinux`, which provides common configuration values and exposes the `features` attribute, as explained below.
```

```suggestion
[`pkgs.linuxManualConfig`](https://github.com/NixOS/nixpkgs/blob/f9d4f5f23ce7ccb7ad91d6d2150e367baf7014f6/pkgs/os-specific/linux/kernel/manual-config.nix) requires a complete configuration to be passed.
It has fewer additional features than `pkgs.buildLinux`, and for example uses common configuration values and exposes the `features` attribute, as explained below.
```

Author: I believe both would be correct, but I went with fewer in this case.

Both functions have an argument `kernelPatches` which should be a list of `{name, patch, extraConfig}` attribute sets, where `name` is the name of the patch (which is included in the kernel’s `meta.description` attribute), `patch` is the patch itself (possibly compressed), and `extraConfig` (optional) is a string specifying extra options to be concatenated to the kernel configuration file (`.config`).

The kernel derivation created with `pkgs.buildLinux` exports an attribute `features` specifying whether optional functionality is or isn’t enabled. This is used in NixOS to implement kernel-specific behaviour.

:::{.example #ex-skip-package-from-kernel-feature}

# Skipping an external package because of a kernel feature

For instance, if the kernel has the `iwlwifi` feature (i.e., has built-in support for Intel wireless chipsets), then NixOS doesn’t have to build the external `iwlwifi` package:

```nix
modulesTree = [kernel]
  ++ pkgs.lib.optional (!kernel.features ? iwlwifi) kernelPackages.iwlwifi
  ++ ...;
```

:::

> If you are using a kernel packaged in Nixpkgs, you can customize it by overriding its arguments. For details on how each argument affects the generated kernel, refer to [the source](https://github.com/NixOS/nixpkgs/blob/f9d4f5f23ce7ccb7ad91d6d2150e367baf7014f6/pkgs/os-specific/linux/kernel/generic.nix) for `pkgs.buildLinux`.

Reviewer: ```suggestion
If you are using a kernel packaged in Nixpkgs, you can customize it by overriding its arguments. For details on how each argument affects the generated kernel, refer to [the `pkgs.buildLinux` source code](https://github.com/NixOS/nixpkgs/blob/f9d4f5f23ce7ccb7ad91d6d2150e367baf7014f6/pkgs/os-specific/linux/kernel/generic.nix).
```

:::{.example #ex-using-linux-manual-config}

# Using `pkgs.linuxManualConfig` with a specific source, version, and config file

```nix
{ pkgs, ... }: {
  version = "6.1.55";
  src = pkgs.fetchurl {
    url = "https://cdn.kernel.org/pub/linux/kernel/v6.x/linux-${version}.tar.xz";
    hash = "sha256:1h0mzx52q9pvdv7rhnvb8g68i7bnlc9rf8gy9qn4alsxq4g28zm8";
  };
  configfile = ./path_to_config_file;
  linux = pkgs.linuxManualConfig {
    inherit version src configfile;
    allowImportFromDerivation = true;
  };
}
```

If necessary, the version string can be slightly modified to explicitly mark it as a custom version. If you do so, ensure the `modDirVersion` attribute matches the source's version, otherwise the build will fail.

```nix
{ pkgs, ... }: {
  version = "6.1.55-custom";
  modDirVersion = "6.1.55";
  src = pkgs.fetchurl {
    url = "https://cdn.kernel.org/pub/linux/kernel/v6.x/linux-${modDirVersion}.tar.xz";
    hash = "sha256:1h0mzx52q9pvdv7rhnvb8g68i7bnlc9rf8gy9qn4alsxq4g28zm8";
  };
  configfile = ./path_to_config_file;
  linux = pkgs.linuxManualConfig {
    inherit version modDirVersion src configfile;
    allowImportFromDerivation = true;
  };
}
```

:::

> Additional attributes can be used with `linuxManualConfig` for further customisation. You're encouraged to read the source for `pkgs.linuxManualConfig` to understand how to use them.

Reviewer: You can now link to the source so people don't have to figure out  how to find it!

Reviewer: Okay, you did above. Maybe not. Ignore this if you want.

Author: Added another link here just to make it easier for the reader.
